### PR TITLE
contrib: clarify new coding standards

### PR DESCRIPTION
**Pull request by sougou**
Signed-off-by: Sugu Sougoumarane <ssougou@gmail.com>

### CODE REVIEW

File: content/en/docs/contributing/code-reviews.md [hunk 1]
```diff
@@ -1,34 +1,49 @@
 ---
-title: Code Reviews
+title: Coding Standards
 ---
 
-Every GitHub pull request must go through a code review and get approved before it will be merged into the master branch.
+## Important rules
+
+Vitess is being used to power many mission-critical production workloads at very large scale. Moreover, many users deploy directly from the master branch. It is extremely important that the changes made by contributrors do not break any existing workloads.
```
Comment (shlomi-noach) [inline content/en/docs/contributing/code-reviews.md]: typo and markdown
```suggestion
Vitess is being used to power many mission-critical production workloads at very large scale. Moreover, many users deploy directly from the `master` branch. It is extremely important that the changes made by contributors do not break any existing workloads.
```

File: content/en/docs/contributing/code-reviews.md [hunk 2]
```diff
@@ -1,34 +1,49 @@
...
+
+In order to avoid disruption, the following concerns need to be kept in mind:
+* Does the change affect any external APIs? If so, make sure that the change satisfies the [compatibility rules](https://github.com/vitessio/enhancements/blob/master/veps/vep-3.md).
+* Can the change introduce a performance regression? If so, it will be good to measure the impact using benchmarks.
```
Comment (shlomi-noach) [inline content/en/docs/contributing/code-reviews.md:13]: This is an issue I've been struggling with. What are the benchmarks? We unfortunately can't tell developers that they should check for performance regression if we don't have the framework to measure that.

File: content/en/docs/contributing/code-reviews.md [hunk 3]
```diff
@@ -1,34 +1,49 @@
...
+* If the change is substantial or is a breaking change, you must publish the proposal as an issue with a title like `RFC: Changing behavior of feature xxx`. Following this, sufficient time has to be given for others to give feedback. A breaking change must still satisfy the compatibility rules.
```
Comment (shlomi-noach) [inline content/en/docs/contributing/code-reviews.md:14]: Suggestion: formalize this even more by adding a `breaking change` label. Not sure if the casual user will have privileges to do that.

As alternative, we should use an issue template, which the developer must use, and which explicitly includes the question "is this a breaking change?"

File: content/en/docs/contributing/code-reviews.md [hunk 4]
```diff
@@ -1,34 +1,49 @@
...
+* New features that affect existing behavior must be introduced "behind a flag". Users will then be encouraged to enable them, but will have the option to fallback to the old behavior if issues are found.
+* New features that do not affect existing behavior have more wiggle room. For example, you do not necessarily have to write tests during your initial commits. But all requirements must eventually be met before announcing the feature as ready.
 
 ## What to look for in a Review
 
-Both authors and reviewers need to answer these general questions:
+Every GitHub pull request must go through a code review and get approved before it will be merged into the master branch.
 
+Every pull request should meet the following requirements:
```
Comment (shlomi-noach) [inline content/en/docs/contributing/code-reviews.md:21]: I'd like to further make a suggestion: pull requests should not come out of the blue. A PR should be associated with an issue. 
e.g. https://github.com/vitessio/vitess/issues/6661 is the tracking issue for the https://github.com/vitessio/vitess/pull/6668 PR. Reason is that an Issue is a better place to discuss the problem in general, the need for a change. A PR is a specific implementation detail. There can be multiple iterative PRs addressing the same issue. So, for example, my own https://github.com/vitessio/vitess/pull/6547 is _not_ a good example, because the discussion is entirely in the PR.

That's a slight overhead for th edeveloper (create an issue, create a PR), but I found that it is also a good preventive mechanism. The developer will open an issue, and I (as a maintainer) can tell them "this is not something that is within the interests of th eproject" or indicate some complexities in the problem, or indicate that there's actually a simple workaround. This may prevent the PR from taking place in the first place, saving time for the developer.

**Review by shlomi-noach — COMMENTED** [review]
Some thoughts and comments inline

File: content/en/docs/contributing/code-reviews.md [hunk 5]
```diff
@@ -1,34 +1,68 @@
...
+
+We use unit tests both to test the code and to describe it for other developers. 
+
+* Unit tests should:
+  * Demonstrate every use case the change covers.
+  * Involve all important units being added or changed.
+  * Attempt to cover every corner case the change introduces. 
+  The thumb rule is: if it can happen in production, it must be covered.
+* Integration tests should ensure that the feature works end-to-end. 
+It must cover all the important use cases of the feature.
```
Comment (deepthi) [inline content/en/docs/contributing/code-reviews.md]:
```suggestion
They must cover all the important use cases of the feature.
```

**Review by systay — APPROVED** [review]
